contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -387,6 +450,126 @@ To deploy a containerized third-party CSI volume driver, it is recommended that
 
 Alternatively, deployment could be simplified by having all components (including external-provisioner and external-attacher) in the same pod (DaemonSet). Doing so, however, would consume more resources, and require a leader election protocol (likely https://git.k8s.io/contrib/election) in the `external-provisioner` and `external-attacher` components.
 
+#### Topology Representation in Node Objects
+Topology information will be represented as labels.
+
+Requirements:
+* Must adhere to the [label format](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+* Must support different drivers on the same node.
+* The format of each key/value pair must match those in `PersistentVolume` and `StorageClass` objects, as described in the [Provisioning and Deleting](#provisioning-and-deleting) section.
+
+Proposal: `"com.example.topology/rack": "rack1"`
+The list of topology keys known to the driver is stored separately in the `CSINodeInfo` object.
+
+Justifications:
+* No strange separators needed, comparing to the alternative. Cleaner format.
+* The same topology key could be used across different sources (different storage plugin, network plugin, etc.)
+* Once NodeRestriction is moved to the newer model (see [here](https://github.com/kubernetes/community/pull/911) for context), for each new label prefix introduced in a new driver, the cluster admin has to configure NodeRestrictions to allow the driver to update labels with the prefix. Cluster installations could include certain prefixes for pre-installed drivers by default. This is less convenient compared to the alternative, which can allow editing of all CSI drivers by default using the “csi.kubernetes.io” prefix, but often times cluster admins have to whitelist those prefixes anyway (for example ‘cloud.google.com’)
+
+Considerations:
+* Upon driver deletion/upgrade/downgrade, stale labels will be left untouched. It’s difficult for the driver to decide whether other sources rely on this label.
+* During driver installation/upgrade/downgrade, controller deployment must be brought down before node deployment, and node deployment must be deployed before the controller deployment, because provisioning relies on up-to-date node information.
```

Comment:
I can't think of an issue immediately. Will update this line to discuss why this isn't an issue and what's the behavior if both deployments are changed in parallel. I have a comment in the Upgrades & Downgrades sections about recommending users to still upgrade those deployments independently, and I'll leave that there just to be safe.

Reply:
One potential issue: if only topology values change while keys remain the same, and if AllowedTopologies is not specified, requisite topology will contain both old and new topology values, and CSI driver may fail the `CreateVolume()` call.

Reply:
I think considering that topology labels can be shared with other components, we cannot override existing topology values if the key already exists on the node.

Reply:
I think users should be free to do so if they can guarantee other components can handle new values too. Maybe we could call this out as a caveat when performing driver upgrades.

Reply:
I would prefer in that case to have the user delete the label manually first instead of us just blindly overriding it and potentially messing up other components.